src/puppetlabs/puppetdb/client.clj

```diff
@@ -6,78 +6,80 @@
             [clj-http.client :as http-client]
             [puppetlabs.puppetdb.command.constants :refer [command-names]]
             [puppetlabs.puppetdb.cheshire :as json]
-            [puppetlabs.kitchensink.core :as kitchensink]))
+            [puppetlabs.puppetdb.schema :refer [defn-validated]]
+            [puppetlabs.puppetdb.utils :as utils]
+            [puppetlabs.kitchensink.core :as kitchensink]
+            [schema.core :as s]))
 
-(defn submit-command-via-http!
+(defn-validated submit-command-via-http!
   "Submits `payload` as a valid command of type `command` and
   `version` to the PuppetDB instance specified by `host` and
   `port`. The `payload` will be converted to JSON before
   submission. Alternately accepts a command-map object (such as those
   returned by `parse-command`). Returns the server response."
-  ([host port command version payload]
-     {:pre [(string? command)
-            (integer? version)]}
+  ([base-url
+    command :- s/Str
+    version :- s/Int
+    payload]
      (->> payload
           (command/assemble-command command version)
-          (submit-command-via-http! host port)))
-  ([host port command-map]
-     {:pre [(string? host)
-            (integer? port)
-            (map? command-map)]}
+          (submit-command-via-http! base-url)))
+  ([base-url :- (s/both utils/base-url-schema
+                        (s/pred #(if-let [v (:version %)]
```

Comment:
Seems like if you moved this pred over to utils as a def, you could name it the same as the description you have below, or something similar and it becomes more intention revealing and more concise in the function definition